cmd/chihaya/main.go

```diff
@@ -65,67 +64,92 @@ func rootCmdRun(cmd *cobra.Command, args []string) error {
 	}
 
 	logic := middleware.NewLogic(cfg.Config, peerStore, preHooks, postHooks)
-	if err != nil {
-		return errors.New("failed to create TrackerLogic: " + err.Error())
-	}
 
-	shutdown := make(chan struct{})
 	errChan := make(chan error)
 
-	var httpFrontend *httpfrontend.Frontend
-	var udpFrontend *udpfrontend.Frontend
+	httpFrontend, udpFrontend := startFrontends(cfg.HTTPConfig, cfg.UDPConfig, logic, errChan)
 
-	if cfg.HTTPConfig.Addr != "" {
-		httpFrontend = httpfrontend.NewFrontend(logic, cfg.HTTPConfig)
+	shutdown := make(chan struct{})
+	quit := make(chan os.Signal)
+	restart := make(chan os.Signal)
+	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
+	signal.Notify(restart, syscall.SIGUSR1)
 
-		go func() {
-			log.Infoln("started serving HTTP on", cfg.HTTPConfig.Addr)
-			if err := httpFrontend.ListenAndServe(); err != nil {
-				errChan <- errors.New("failed to cleanly shutdown HTTP frontend: " + err.Error())
+	go func() {
+		for {
+			select {
+			case <-restart:
+				log.Info("Got signal to restart")
+
+				// Reload config
+				configFile, err = ParseConfigFile(configFilePath)
+				if err != nil {
+					log.Error("failed to read config: " + err.Error())
+					continue
+				}
+				cfg = configFile.MainConfigBlock
+
+				preHooks, postHooks, err = configFile.CreateHooks()
+				if err != nil {
+					log.Error("failed to create hooks: " + err.Error())
+					continue
+				}
+
+				// Stop frontends
+				log.Debug("Stopping frontends")
+				if udpFrontend != nil {
+					udpFrontend.Stop()
+				}
+
+				if httpFrontend != nil {
+					httpFrontend.Stop()
+				}
+
+				// Stop hooks
+				log.Debug("Stopping logic")
+				errs := logic.Stop()
+				for _, err := range errs {
+					errChan <- err
+				}
+
+				// Restart
+				log.Debug("Restarting logic")
+				logic = middleware.NewLogic(cfg.Config, peerStore, preHooks, postHooks)
+
+				log.Debug("Restarting frontends")
+				httpFrontend, udpFrontend = startFrontends(cfg.HTTPConfig, cfg.UDPConfig, logic, errChan)
+
+				log.Debug("Successfully restarted")
+				continue
+
+			case <-quit:
+			case <-shutdown:
 			}
-		}()
-	}
-
-	if cfg.UDPConfig.Addr != "" {
-		udpFrontend = udpfrontend.NewFrontend(logic, cfg.UDPConfig)
 
-		go func() {
-			log.Infoln("started serving UDP on", cfg.UDPConfig.Addr)
-			if err := udpFrontend.ListenAndServe(); err != nil {
-				errChan <- errors.New("failed to cleanly shutdown UDP frontend: " + err.Error())
+			// Stop frontends
```

Comment:
I think moving this into the `case` block for _shutdown_ and then explicitly adding `fallthrough` inside the _quit_ case makes the logic more clear.